Library of Congress authorities

mbklein opened the pull request:
* Create generic, `use`-able `LOC.Base module` from the existing `LCNAF` module
* Create `Names`, `SubjectHeadings`, and `Languages` authorities from `LOC.Base`
* Create generic `Authoritex.TestCase` to test the idea of shared specs
* Create test modules for all LOC implementations using `Authoritex.TestCase`
* Move discovery of authorities from module attribute to config

I also changed the top-level Authoritex test to use its own cassettes because sharing the LCNAF cassettes became cumbersome.

kdid commented:
Just a couple things to handle better so we can make sure we're providing enough helpful error info to users. 

```
iex(1)> Authoritex.fetch("info:lc/authorities/sh93009534")
{:error, 303}

iex(15)> Authoritex.fetch("http://id.loc.gov/authorities/subjects")          
** (Protocol.UndefinedError) protocol Enumerable not implemented for nil of type Atom. This protocol is implemented for the following type(s): HashSet, Range, Map, Function, List, Stream, Date.Range, HashDict, GenEvent.Stream, MapSet, File.Stream, IO.Stream
    (elixir) lib/enum.ex:1: Enumerable.impl_for!/1
    (elixir) lib/enum.ex:141: Enumerable.reduce/3
    (elixir) lib/stream.ex:952: Stream.do_enum_transform/7


iex(22)> Authoritex.search("lcnaf", "")   
{:error, %HTTPoison.Error{id: nil, reason: :timeout}}

```

mbklein commented:
> ```
> iex(1)> Authoritex.fetch("info:lc/authorities/sh93009534")
> {:error, 303}
> ```

Fixed to handle `3xx` redirects:
```
iex(1)> Authoritex.fetch("info:lc/authorities/sh93009534")
{:ok, "Tibbett, Henry (Fictitious character)"}
```

> ```
> iex(15)> Authoritex.fetch("http://id.loc.gov/authorities/subjects")          
> ** (Protocol.UndefinedError) protocol Enumerable not implemented for nil of type Atom. This protocol is implemented for the following type(s): HashSet, Range, Map, Function, List, Stream, Date.Range, HashDict, GenEvent.Stream, MapSet, File.Stream, IO.Stream
>     (elixir) lib/enum.ex:1: Enumerable.impl_for!/1
>     (elixir) lib/enum.ex:141: Enumerable.reduce/3
>     (elixir) lib/stream.ex:952: Stream.do_enum_transform/7
> ```

Valid-looking IDs with responses the parser doesn't expect to see will now return and error and the unexpected content. This is true whether the response is HTML, valid XML or even valid RDFXML, or anything else. If the fetch parser doesn't like it, it's a `:bad_response`.
```
iex(2)> Authoritex.fetch("http://id.loc.gov/authorities/subjects")
{:error, :bad_response, "<rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">..."}}
```

> ```
> iex(22)> Authoritex.search("lcnaf", "")   
> {:error, %HTTPoison.Error{id: nil, reason: :timeout}}
> ```

This is valid search and a real timeout. If I run it a second time, I get a valid list of results. We should probably use a timeout library like [HTTPoison.Retry](https://hexdocs.pm/httpoison_retry/HTTPoison.Retry.html) to make retries automatic, but for now, this is close to the best we've got.

mbklein commented:
I also removed the 500 tests from the shared specs and moved them to the `LOC` test module. My reasoning:

- Keeping multiple custom cassettes up to date for what amounts to the same test for all of them is a huge pain
- No two services are going to respond the same way under error conditions, so this is kind of a meaningless test anyway
- We can always write service-specific error tests into the service's test module (like I did with this round of LOC changes)
